lib/Onyx.js

```diff
@@ -853,26 +909,17 @@ function set(key, value) {
         return remove(key);
     }
 
-    // eslint-disable-next-line no-use-before-define
-    if (hasPendingMergeForKey(key)) {
-        Logger.logAlert(`Onyx.set() called after Onyx.merge() for key: ${key}. It is recommended to use set() or merge() not both.`);
-    }
+    // We relay changes to subscribers optimistically to avoid lag (think UI updates)
+    notify(key, value, 'set');
 
-    // If the value in the cache is the same as what we have then do not update subscribers unless they
-    // have initWithStoredValues: false then they MUST get all updates even if nothing has changed.
-    if (!cache.hasValueChanged(key, value)) {
-        cache.addToAccessedKeys(key);
-        notifySubscribersOnNextTick(key, value, subscriber => subscriber.initWithStoredValues === false);
-        return Promise.resolve();
-    }
-
-    // Adds the key to cache when it's not available
-    cache.set(key, value);
-    notifySubscribersOnNextTick(key, value);
+    const nextModification = () => Storage.setItem(key, value)
+        .catch((error) => {
+            // We have to resolve the current promise for the key, so that the next set doesn't block
+            keyOperations[key] = Promise.resolve();
```

Comment:
Yes, as @kidroca wrote, this is because the "inner" `set` call in `evictStorageAndRetry` would be blocked by the "outer" `set` call, because there will already be a promise in the `keyOperations`.

I don't know what the expected behaviour is, if `Storage.setItem` rejects after evicting. Should it try to evict again or just abort? If eviction should happen again, we will need to pass `set` rather than `Storage.setItem`